pkg/util/ring/adaptive_chan.go

```diff
@@ -0,0 +1,285 @@
+// intended usage pattern:
+//
+//  1. Create an adaptive channel.
+//  2. Provide the send-only channel to your producer(s). They are responsible
+//     for closing this channel when they're done. If more than one goroutine
+//     will have access to this channel, then it's the producer's responsibility
+//     to coordinate the channel close operation.
+//  3. Provide the receive-only channel to your consumer(s), and let them
+//     receive from it with the two return value syntax for channels in order to
+//     check for termination from the sending side.
+//  4. Use the AdaptiveChanController to control the internal buffer behaviour
+//     and to monitor stats. This should typically be held by the creator of the
+//     adaptive channel. Refrain from holding a reference to the send-only
+//     channel to force termination of the producing side. Instead, provide a
+//     side mechanism to communicate the intention of terminating the sending
+//     side, e.g. providing your producer(s) with a context as well as the
+//     send-only channel. An adaptive channel is meant as a queueing system, not
+//     as a coordination mechanism for producer(s), consumer(s) and
+//     controller(s).
+//
+// This pattern is designed to maximize decoupling while providing insights and
+// granular control on memory usage. While the controller is not meant to make
+// any direct changes to the queued data, the Clear method provides the
+// opportunity to discard all queued items as an administrative measure. This
+// doesn't terminate the queue, though, i.e. it doesn't close the send-only
+// channel.
+func AdaptiveChan[T any]() (send chan<- T, recv <-chan T, ctrl *AdaptiveChanController) {
+	internalSend := make(chan T)
+	internalRecv := make(chan T)
+	statsChan := make(chan AdaptiveChanStats)
+	cmdChan := make(chan acCmd)
+	ctrl = &AdaptiveChanController{
+		statsChan: statsChan,
+		cmdChan:   cmdChan,
+	}
+
+	go func() {
+		defer close(internalRecv)
+		defer close(statsChan)
+
+		var q Ring[T]
+		var stats AdaptiveChanStats
+
+		// the loop condition is that we either have items to dequeue or that we
+		// have the possibility to receive new items to be queued
+		for stats.Len > 0 || internalSend != nil {
+			// NOTE: the overhead of writing stats in each iteration is
+			// negligible. I tried a two phase stats writing with a chan
+			// struct{} to get notified that the controller wanted stats, then
+			// updating the stats and finally writing to statsChan. There was no
+			// observable difference for just enqueueing and dequeueing after
+			// running the benchmarks several times, and reading stats got worse
+			// by ~22%
+			q.WriteStats(&stats.RingStats)
+
+			// if we don't have anything in the queue, then make the dequeueing
+			// branch of the select block indefinitely by providing a nil
+			// channel, leaving only the queueing branch available
+			dequeueChan := internalRecv
+			if stats.Len == 0 {
```

Comment:
Hmm, in this case I think it's fine because it's not `*Ring`'s internal state, but rather the adaptive channel's internal state. IMO, if we didn't need any other stat, I would say calling `q.WriteStats` and reading `Len` is definitely unnecessary, and I would say we should just use `q.Len()` in that case. On the other hand, as we already call `q.WriteStats` on every iteration and it's always up to date, then I would say the contrary, I think that calling `q.Len` is unnecessary as we have already asked for that information while calling `q.WriteStats` as well. And with "unnecessary" I don't mean it would be even noticeable, but rather that I tend to rely on the facts I have just gathered and that I know are fresh.

In any case, I think this would be more stylistic and I don't have a strong feeling for it, I can change it for `q.Len` if you prefer.

Reply:
I think the code reads a little more cleanly if we don't use the stats for anything operational and keep them for reporting purposes only.